sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -566,26 +568,73 @@ cdef class PairwiseDistancesArgKmin(PairwiseDistancesReduction):
+              synchronisation.
+
+              - 'auto' relies on a simple heuristic to choose between
+              'parallel_on__X' and 'parallel_on_Y'.
+
+              - None (default) looks-up in scikit-learn configuration for
+              `pairwise_dist_parallel_strategy`, and use 'auto' if it is not set.
+
+        return_distance : boolean, default=False
+            Return distances between each X vector and its
+            argkmin if set to True.
+
         Returns
         -------
-        argkmin: PairwiseDistancesArgKmin
-            The suited PairwiseDistancesArgKmin implementation.
+            Indices of argkmin for each vector in X and its associated distances
+            if return_distance=True.
+
+        Notes
+        -----
+            This public classmethod is responsible for introspecting the arguments
+            values to dispatch to the private :meth:`PairwiseDistancesArgKmin._compute`
+            instance method of the most appropriate :class:`PairwiseDistancesArgKmin`
+            concrete implementation.
+
+            All temporarily allocated datastructures necessary for the concrete
+            implementation are therefore freed when this classmethod returns.
+
+            This allows entirely decoupling the interface entirely from the
+            implementation details whilst maintaining RAII.
         """
-        # This factory comes to handle specialisations.
-        if metric in ("fast_euclidean", "fast_sqeuclidean") and not issparse(X) and not issparse(Y):
+        # Note (jjerphan): Some design thoughts for future extensions.
+        # This factory comes to handle specialisations for the given arguments.
+        # For future work, this might can be an entrypoint to specialise operations
+        # for various back-end and/or hardware and/or datatypes, and/or fused
+        # {sparse, dense}-datasetspair etc.
+        if (
+            metric in ("fast_euclidean", "fast_sqeuclidean")
+                and not issparse(X)
+                and not issparse(Y)
+        ):
             use_squared_distances = metric == "fast_sqeuclidean"
-            return FastEuclideanPairwiseDistancesArgKmin(
+            pda = FastEuclideanPairwiseDistancesArgKmin(
                 X=X, Y=Y, k=k,
                 use_squared_distances=use_squared_distances,
                 chunk_size=chunk_size,
                 metric_kwargs=metric_kwargs,
             )
+        else: # Fall back on the default
+            pda = PairwiseDistancesArgKmin(
+                datasets_pair=DatasetsPair.get_for(X, Y, metric, metric_kwargs),
+                k=k,
+                chunk_size=chunk_size,
+            )
 
-        return PairwiseDistancesArgKmin(
-            datasets_pair=DatasetsPair.get_for(X, Y, metric, metric_kwargs),
-            k=k,
-            chunk_size=chunk_size,
-        )
+        return pda._compute(strategy=strategy, return_distance=return_distance)
```

Comment:
I think your points regarding distinguishing allocation and de-allocation for `parallel_on_X` or `parallel_on_Y` makes sense and we should find a way to implement this.

To clarify to me, private attributes allows having generic interfaces (`_parallel_on_{X|Y}` and friends) which are decoupled from the actual data-structures which are being manipulated by each concrete `PairwiseDistancesReduction`.
For the sake of explicitness, `PairwiseDistancesArgKmin` modifies a structure of two array (namely `(argkmin_indices, argkmin_distances)`) but the subsequent `PairwiseDistancesReduction` might modify other kind of data-structures.

This makes me think that I should precise this point of design somewhere, because this has been and is somewhat implicit now.

I would also be happy discussing it again on a call!

Reply:
@thomasjpfan: I have been wrapping my head around your comment's suggestion in your message above:

> [...] maybe even malloc in __init__ `heaps_r_distances_chunks[n_threads]` for the "parallel_on_Y" case, if there are no performance issues

and I have a naive question: why would this be useful to allocate them in `__init__` exactly? [Their current allocation allow making use of data locality.](https://github.com/jjerphan/scikit-learn/blob/pairwise-distances-argkmin-raii/sklearn/metrics/_pairwise_distances_reduction.pyx#L744-L757) in `parallel_on_Y`.

On a side note, the eventual  caveat that I (as of my understanding of your message) see is having the buffers pointed by `heaps_r_distances_chunks[i]` and `heaps_indices_chunks[i]` (for i in {0, ..., n_thread -1}) not be de-allocated -- IIRC, we briefly have imagined this case after you mentioned it in a call.

I think this can append if the following conditions are verified:
 - `parallel_on_Y` fails to complete (for some reason), not calling `_parallel_on_Y_finalize` and thus not [freeing those structure there](https://github.com/jjerphan/scikit-learn/blob/pairwise-distances-argkmin-raii/sklearn/metrics/_pairwise_distances_reduction.pyx#L800-L803)
 - the `PairwiseDistancesArgKmin` instance is de-allocated after this failure (e.g. when the caller returns)
 - freeing a buffer of pointers does not free the allocated data-structures allocated on those pointers (i.e. no cascading when freeing), which is the normal behavior of `free`.

A simple solution to prevent this scenario from happening would be to extend [`__dealloc__`](https://github.com/jjerphan/scikit-learn/blob/pairwise-distances-argkmin-raii/sklearn/metrics/_pairwise_distances_reduction.pyx#L672-L677) to free `heaps_r_distances_chunks[i]` and `heaps_indices_chunks[i]` (for i in {0, ..., n_thread -1}) if not already free before freeing `heaps_r_distances_chunks` and `heaps_indices_chunks`.

What do you think of it? Is this addition needed? Is this sufficient for this small PR?

Reply:
> why would this be useful to allocate them in __init__ exactly? Their current allocation allows making use of data locality. in parallel_on_Y.

There would be more symmetry between `__dealloc__` and `__(c)init__`. I was thinking of calling `_parallel_on_Y_parallel_init` during `__(c)init__` if the strategy is `parallel_on_Y`.

> A simple solution to prevent this scenario from happening would be to extend __dealloc__ to free heaps_r_distances_chunks[i] and heaps_indices_chunks[i]

I think that should work. This means the strategy needs to be determined during `__init__`. Then `__dealloc__` does the correct dealloc when the strategy is `parallel_on_Y`.

Reply:
> There would be more symmetry between __dealloc__ and __(c)init__. I was thinking of calling _parallel_on_Y_parallel_init during __(c)init__ if the strategy is parallel_on_Y.

Ah yes, pairing allocation and deallocation where they are naturally meant to be to would indeed be better.

> I think that should work. This means the strategy needs to be determined during __init__. Then __dealloc__ does the correct dealloc when the strategy is parallel_on_Y.

7b6b39925c4a93a6074bc148b5324b29f6e64bcc put the choice of the strategy in `__init__`.
Please, let me know if this correctly works for you.

To add to your remark, I think the current implementation might be OK and "symmetric":
 - the pointers are allocated in `__init__`
 - for `parallel_on_Y`, the heap-allocated buffers are allocated at the start within `_parallel_on_Y_parallel_init`
 - for `parallel_on_Y`, those heap-allocated buffers are deallocated at the end within `_parallel_on_Y_finalize`
 - the pointers are deallocated in and `__dealloc__`

I am working on moving within `__init__` and `__dealloc__`. I think this might make the code a bit harder to follow and might cause some performances regression. Let's see if this can be made possible.

Reply:
In fact, I think sticking to the implementations (as of a6ab85f771db7e2d28d8c03c273712dbcb65b6da) might better/simpler/easier. Here are the pros that I currently see:
 - `_parallel_on_Y_parallel_init` and `_parallel_on_Y_finalize` are overloaded by `FastEuclideanPairwiseDistancesArgKmin`: moving their logic in `__init__` and `__dealloc__` might make the overall logic a bit more complex because we would need to extend `FastEuclideanPairwiseDistancesArgKmin.{__init__,__dealloc__}`. This is my opinion, and I would accept others'.
 - this pattern (using `_parallel_on_Y_parallel_init` and `_parallel_on_Y_finalize`) allows using as little parallel section as possible and making sure data locality is preserved; which is the best regarding performances as explains by [this piece of documentation](https://docs.oracle.com/cd/E19059-01/stud.10/819-0501/7_tuning.html).

Reply:
I see what you mean. Okay, let's go with what we have now. One really needs to be careful with `_parallel_on_Y_parallel_init` + `_parallel_on_Y_finalize`.

Reply:
I agree. Potentially, we could rework that later if we find another pattern. For instance (and with gut feeling), `compute` could have everything done straight in an instance's `__cinit__` and in `__dealloc__`: in this case, results could be attached to the instance as attributes and could simply be returned at the end of `compute`. I think this would allow having a single parallel section.

In any cases, I am waiting for your explicit LGTM before merging in.